File: apps/mobile-ops/modules/youverify-liveness/ios/YouVerifyLivenessModule.swift
import ExpoModulesCore
import AzureAIVisionFaceUI   // CocoaPod: pod 'AzureAIVisionFaceUI', '~> 0.19.0-beta'
import SwiftUI
import UIKit

// ---------------------------------------------------------------------------
// YouVerifyLivenessModule
//
// Wraps the Azure AI Vision Face Liveness Detection SDK so that React Native
// can launch a native face liveness challenge with a single JS call.
//
// The sessionToken is an Azure Face API JWT issued by YouVerify's backend
// (POST /v2/api/identity/sdk/liveness/token). It is passed directly to the
// Azure Face Liveness SDK — no additional credential handling needed here.
//
// Requires iOS 15+ (set in Podfile / Deployment Target).
// Requires NSCameraUsageDescription in Info.plist.
// ---------------------------------------------------------------------------

public class YouVerifyLivenessModule: Module {
  private var hasCompleted = false

  public func definition() -> ModuleDefinition {
    Name("YouVerifyLiveness")

    AsyncFunction("startLiveness") { (options: [String: Any], promise: Promise) in
      guard
        let sessionId = options["sessionId"] as? String, !sessionId.isEmpty,
        let sessionToken = options["sessionToken"] as? String, !sessionToken.isEmpty
      else {
        promise.reject("INVALID_INPUT", "sessionId and sessionToken are required")
        return
      }

      let sandbox = options["sandbox"] as? Bool ?? true

      DispatchQueue.main.async {
        self.launchLiveness(
          sessionId: sessionId,
          sessionToken: sessionToken,
          sandbox: sandbox,
          promise: promise
        )
      }
    }
  }

  // ── Private helpers ──────────────────────────────────────────────────────

  private func launchLiveness(
    sessionId: String,
    sessionToken: String,
    sandbox: Bool,
    promise: Promise
  ) {
    hasCompleted = false

    guard let rootVC = currentRootViewController() else {
      promise.reject("NO_ROOT_VC", "Unable to find root view controller")
      return
    }

    // Build the Azure Face Liveness session configuration.
    // AzureAIVisionFaceUI.FaceAnalyzerCreateSession is the entry point.
    // Reference: https://github.com/Azure-Samples/azure-ai-vision-sdk
    let config = VisionServiceOptions()
    // The sessionToken is the Azure Face API auth token.
    // It already encodes the endpoint and session constraints.
    let sessionOptions = FaceSessionOptions()
    sessionOptions.setAuthToken(sessionToken)

    // Create and start the liveness session.
    FaceAnalyzer.create(
      serviceOptions: config,
      input: .cameraInput,
      completionHandler: { [weak self] faceAnalyzer, error in
        if let error = error {
          promise.reject("SDK_INIT_FAILED", error.localizedDescription)
          return
        }

        guard let analyzer = faceAnalyzer else {
          promise.reject("SDK_INIT_FAILED", "FaceAnalyzer could not be created")
          return
        }

        // FaceAnalyzerView is the SwiftUI/UIKit view that hosts the camera + challenge.
        let faceView = FaceAnalyzerView(analyzer: analyzer)
        let hostingVC = UIHostingController(rootView: faceView)
        hostingVC.modalPresentationStyle = .fullScreen

        // Observe the liveness result.
        analyzer.resultReceived = { result in
          DispatchQueue.main.async {
            guard !self.hasCompleted else {
              return
            }
            self.hasCompleted = true
            hostingVC.dismiss(animated: true) {
              let passed = result.livenessDecision == .realFace
              let faceImageB64 = result.detectedFaces?.first?.boundingBox != nil
                ? self?.extractFaceImage(from: result)
                : nil

              let output: [String: Any?] = [
                "passed": passed,
                "faceImageB64": faceImageB64,
                "sessionId": sessionId,
                "errorCode": passed ? nil : "LIVENESS_FAILED",
                "errorMessage": passed ? nil : "Liveness check did not pass",
              ]
              promise.resolve(output.compactMapValues { $0 })
            }
          }
        }

        // Handle session errors (expired token, network, camera denied, etc.).
        analyzer.sessionCompleted = { session, error in
          if let error = error {
            DispatchQueue.main.async {
              guard !self.hasCompleted else {
                return
              }
              self.hasCompleted = true
              hostingVC.dismiss(animated: true) {
                let nsError = error as NSError
                promise.reject(
                  nsError.domain,
                  error.localizedDescription
                )
              }
            }
          }
        }

        rootVC.present(hostingVC, animated: true)

        // Start the liveness session using the Azure Face token.
        do {
          try analyzer.startAnalyzing(
            usingSessionAuthorizationToken: sessionToken
          )
        } catch {
          hostingVC.dismiss(animated: true)
          promise.reject("SESSION_START_FAILED", error.localizedDescription)
        }
      }
    )
  }

  private func extractFaceImage(from result: FaceAnalysisResult) -> String? {
    guard
      let imageData = result.detectedFaces?.first?.largestFace?.image?.pngData()
    else {
      return nil
    }
    return imageData.base64EncodedString()
  }

  private func currentRootViewController() -> UIViewController? {
    let scene = UIApplication.shared.connectedScenes
      .compactMap { $0 as? UIWindowScene }
      .first { $0.activationState == .foregroundActive }

    let root = scene?.windows.first(where: \.isKeyWindow)?.rootViewController
    return topViewController(from: root)
  }

  private func topViewController(from root: UIViewController?) -> UIViewController? {
    if let navigationController = root as? UINavigationController {
      return topViewController(from: navigationController.visibleViewController)
    }
    if let tabBarController = root as? UITabBarController {
      return topViewController(from: tabBarController.selectedViewController)
    }
    if let presented = root?.presentedViewController {
      return topViewController(from: presented)
    }
    return root
  }
}

File: apps/mobile-ops/modules/youverify-liveness/android/src/main/java/expo/modules/youverifyliveness/YouVerifyLivenessModule.kt
package expo.modules.youverifyliveness

import android.app.Activity
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

// Azure AI Vision Face UI SDK
// Gradle: implementation 'com.azure.android:azure-ai-vision-face-ui:0.19.0-beta.1'
// Repository: https://github.com/Azure-Samples/azure-ai-vision-sdk/tree/main/samples/kotlin/face/FaceAnalyzerSample
import com.azure.android.ai.vision.face.ui.FaceAnalyzer
import com.azure.android.ai.vision.face.ui.FaceAnalyzerCreateOptions
import com.azure.android.ai.vision.face.ui.LivenessFailureReason
import com.azure.android.ai.vision.face.ui.LivenessOperationMode

// ─────────────────────────────────────────────────────────────────────────────
// YouVerifyLivenessModule
//
// Android native module that wraps the Azure AI Vision Face Liveness Detection
// SDK to provide real provider-backed face liveness for the mobile app.
//
// The sessionToken is an Azure Face API JWT issued by YouVerify's backend
// (POST /v2/api/identity/sdk/liveness/token). It authorises one liveness
// session against the Azure Face resource in YouVerify's account.
//
// Requires:
//   - CAMERA permission in AndroidManifest.xml
//   - minSdkVersion 24 (Android 7.0)
// ─────────────────────────────────────────────────────────────────────────────

private const val LIVENESS_REQUEST_CODE = 0x4C56   // "LV" in ASCII hex

class YouVerifyLivenessModule : Module() {

  private var pendingPromise: Promise? = null
  private var pendingSessionId: String? = null

  override fun definition() = ModuleDefinition {
    Name("YouVerifyLiveness")

    AsyncFunction("startLiveness") { options: Map<String, Any>, promise: Promise ->
      if (pendingPromise != null) {
        promise.reject("LIVENESS_ALREADY_RUNNING", "A liveness session is already in progress", null)
        return@AsyncFunction
      }

      val sessionId = options["sessionId"] as? String
      val sessionToken = options["sessionToken"] as? String

      if (sessionId.isNullOrBlank() || sessionToken.isNullOrBlank()) {
        promise.reject("INVALID_INPUT", "sessionId and sessionToken are required", null)
        return@AsyncFunction
      }

      val activity = appContext.activityProvider?.currentActivity
      if (activity == null) {
        promise.reject("NO_ACTIVITY", "No foreground activity available", null)
        return@AsyncFunction
      }

      pendingPromise = promise
      pendingSessionId = sessionId

      try {
        launchLiveness(activity, sessionId, sessionToken)
      } catch (error: Throwable) {
        pendingPromise = null
        pendingSessionId = null
        promise.reject("SESSION_START_FAILED", error.message ?: "Unable to start face verification", error)
      }
    }

    // Wire up Activity result so the module can capture the liveness outcome.
    OnActivityResult { _, payload ->
      if (payload.requestCode != LIVENESS_REQUEST_CODE) return@OnActivityResult

      val promise = pendingPromise ?: return@OnActivityResult
      val sessionId = pendingSessionId ?: ""
      pendingPromise = null
      pendingSessionId = null

      when (payload.resultCode) {
        Activity.RESULT_OK -> {
          val data = payload.data
          val passed = data?.getBooleanExtra(FaceAnalyzer.EXTRA_RESULT_LIVENESS_PASSED, false) ?: false
          val faceImageB64 = data?.getStringExtra(FaceAnalyzer.EXTRA_RESULT_FACE_IMAGE_BASE64)

          promise.resolve(
            mapOf(
              "passed" to passed,
              "faceImageB64" to faceImageB64,
              "sessionId" to sessionId,
            )
          )
        }

        Activity.RESULT_CANCELED -> {
          promise.resolve(
            mapOf(
              "passed" to false,
              "sessionId" to sessionId,
              "errorCode" to "USER_CANCELLED",
              "errorMessage" to "User cancelled the liveness check",
            )
          )
        }

        else -> {
          val data = payload.data
          val code = data?.getStringExtra(FaceAnalyzer.EXTRA_RESULT_FAILURE_REASON)
            ?: "LIVENESS_FAILED"
          val message = failureCodeToMessage(code)

          promise.resolve(
            mapOf(
              "passed" to false,
              "sessionId" to sessionId,
              "errorCode" to code,
              "errorMessage" to message,
            )
          )
        }
      }
    }
  }

  // ── Private helpers ──────────────────────────────────────────────────────

  private fun launchLiveness(activity: Activity, sessionId: String, sessionToken: String) {
    // FaceAnalyzer.createForPassiveLiveness launches a fullscreen Activity that
    // runs the Azure Face Liveness passive challenge using the device camera.
    // The auth token (Azure Face JWT from YouVerify) is passed directly.
    val createOptions = FaceAnalyzerCreateOptions().apply {
      livenessOperationMode = LivenessOperationMode.PASSIVE
    }

    FaceAnalyzer.createForPassiveLiveness(
      activity,
      sessionToken,          // Azure Face JWT — authorises one session
      createOptions,
      LIVENESS_REQUEST_CODE,
    )
  }

  private fun failureCodeToMessage(code: String): String = when (code) {
    LivenessFailureReason.FACE_NOT_DETECTED.name -> "No face detected. Ensure your face is in the frame."
    LivenessFailureReason.FACE_MOVED -> "Face moved during capture. Please hold still."
    LivenessFailureReason.SPOOF_DETECTED.name -> "Liveness check failed."
    LivenessFailureReason.SESSION_EXPIRED.name -> "Session expired. Please try again."
    else -> "Liveness check did not pass."
  }
}
